contrib/go-chi/chi/chitrace.go

```diff
@@ -0,0 +1,66 @@
+// Package chi provides tracing middleware for the Chi web framework.
+package chi
+
+import (
+	"context"
+	"net/http"
+	"strconv"
+	"strings"
+
+	dd "github.com/DataDog/dd-trace-go/opentracing"
+	"github.com/go-chi/chi"
+	"github.com/go-chi/chi/middleware"
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+)
+
+func Middleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Extract span from incoming request if possible, create a new span otherwise.
+		var span opentracing.Span
+		var ctx context.Context
+		if clientSpanCtx, err := opentracing.GlobalTracer().Extract(opentracing.HTTPHeaders, opentracing.HTTPHeadersCarrier(r.Header)); err == nil {
+			span = opentracing.StartSpan("http.request", ext.RPCServerOption(clientSpanCtx))
+			ctx = opentracing.ContextWithSpan(r.Context(), span)
+		} else {
+			span, ctx = opentracing.StartSpanFromContext(r.Context(), "http.request")
+		}
+		defer span.Finish()
+
+		// Pass the span through the request's context
+		rr := r.WithContext(ctx)
+
+		// Wrap the ResponseWriter in order to get access to the status code.
+		ww := middleware.NewWrapResponseWriter(w, r.ProtoMajor)
+
+		// Serve Request
+		next.ServeHTTP(ww, rr)
+
+		var route string
+		// Get the resource associated to this request
+		// It's recommended by Chi to do this after having served the request
+		// https://github.com/go-chi/chi/blob/f2aed534ddaae0ec331ea305ef9a804b67b672f6/context.go#L80
+		rc := chi.RouteContext(rr.Context())
+		if rc == nil {
+			route = "unknown"
+		} else {
+			route = strings.Replace(strings.Join(rc.RoutePatterns, ""), "/*/", "/", -1)
+			if route == "" {
+				route = "unknown"
+			}
+		}
+
+		resource := rr.Method + " " + route
+
+		span.SetTag(dd.SpanType, "http")
+		span.SetTag(string(ext.HTTPMethod), rr.Method)
```

Comment:
You don't need to type cast any of these keys. They are already strings.

Reply:
If I don't I get 
> cannot use ext.HTTPUrl (type ext.stringTagName) as type string in argument to span.SetTag

or am I missing something?

Reply:
Oh, I thought you were using our `ext` package. If you're gonna use that, then shouldn't it be:
```go
ext.HTTPMethod.Set(span, rr.Method)
```
And so on for the rest... ?

Reply:
I'm using opentracing's tags wherever possible, but I'm fine moving all of them do datadog's.

`ext.HTTPMethod.Set(span, rr.Method)` and `span.SetTag(string(ext.HTTPMethod), rr.Method)` are equivalent, IIRC.

Reply:
You are right about those being equivalent, it's just that the first version seems more natural and how the API was intended to be used, as opposed to casting to string.

Either way, I like your suggestion about moving them to Datadog's (using our own ext package). If you don't mind doing that, that would be awesome!

Thoughts?

Reply:
Moving them to Datadog's sounds good. I'll do that next week.

Reply:
Thanks @quentinms ! Appreciate the help.